[LSP] Add semantic token caching for solution open

allisonchou opened the pull request:
- In local C#/VB, semantic tokens are cached in a SQL database that is queried on solution open. This allows us to instantly colorize the solution without performing any token computations. The goal of this PR is to accomplish something similar in LSP scenarios.
- This PR adds the caching described above to Roslyn LSP. It also makes it so we now only return semantic classifications, whereas before we returned both semantic + syntactic classifications. We are able to do this because the C# syntactic classifier runs on the client, so we save ourselves lots of work by only returning semantic classifications.
- We have to special case Razor since Razor requires us to return both syntactic and semantic classifications. This is due to the C# syntactic classifier not running on the client in Razor scenarios. Ideally, Razor would eventually run the classifier in some way (tracked by https://github.com/dotnet/razor-tooling/issues/5850) so we can get rid of this special casing. For now, we cache and return both types of tokens.
- Updated + added tests.

allisonchou commented on src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs:
```diff
@@ -147,8 +148,8 @@ internal static class SemanticClassificationUtilities
-                    await AddSemanticClassificationsAsync(
-                        document, snapshotSpan.Span.ToTextSpan(), classificationService, classifiedSpans, cancellationToken: cancellationToken).ConfigureAwait(false);
+                    await SemanticClassificationCacheUtilities.AddSemanticClassificationsAsync(
+                        document, snapshotSpan.Span.ToTextSpan(), classificationService, classifiedSpans, isRazorDoc: false, cancellationToken).ConfigureAwait(false);
```
The `isRazorDoc` parameter was added for testing purposes so we can compare returning only semantic tokens vs. returning both semantic and syntactic tokens.

allisonchou commented on src/EditorFeatures/Core/Implementation/Classification/SemanticClassificationUtilities.cs:
```diff
@@ -165,56 +166,5 @@ internal static class SemanticClassificationUtilities
             }
         }
-
-        private static async Task AddSemanticClassificationsAsync(
```
This method and the method below were moved to `SemanticClassificationCacheUtilities.cs` so that they're accessible in the LSP layer.

allisonchou commented on src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs:
```diff
@@ -30,5 +36,71 @@ internal static class SemanticClassificationCacheUtilities
+            if (await TryAddSemanticClassificationsFromCacheAsync(document, textSpan, classifiedSpans, isFullyLoaded, cancellationToken).ConfigureAwait(false))
+                return;
+
+            // We need to special case Razor. Since the C# syntactic classifier doesn't run on their end, we need to
```
This if-check and the added `isRazorDoc` parameter are the only new parts of these methods; they were otherwise moved untouched from `SemanticClassificationUtilities.cs`. As previously mentioned, the `isRazorDoc` parameter was added for testing purposes.

allisonchou commented on src/Workspaces/Remote/ServiceHub/Services/SemanticClassificationCache/RemoteSemanticClassificationCacheService.cs:
```diff
@@ -97,11 +97,22 @@ private static async Task CacheSemanticClassificationsAsync(Document document, C
+            // classifications since the client doesn't run the C# classifier in Razor LSP scenarios.
+            // Ideally, Razor will eventually run the C# syntactic classifier on their end so we can remove this
+            // special casing: https://github.com/dotnet/razor-tooling/issues/5850
+            if (document.FilePath is not null && document.FilePath.EndsWith(".g.cs"))
```
`.IsRazorDocument()` doesn't work in OOP scenarios, so as a workaround I added a check for `.EndsWith(".g.cs")` instead which seems to do the trick

CyrusNajmabadi replied:
@NTaylorMullen for thoughts.

CyrusNajmabadi replied:
note: This feels weird as well.  somethign about the entire razor space here seems somewhat broken (necessitating these extra strange codepaths at the feature level).  my preference would be that if razor is *currently* doing something different/weird, that we just not have this optimization/feature for them.  and, if we can then get razor docs/scenarios to work like normal scenarios, then we reenable, with no specialized code *in the feature* to handle razor.

allisonchou replied:
Most of this special casing is due to Razor not running the C# syntax classifier on their end. However, I believe adding support is a pretty big work item and could possibly also negatively impact Razor perf, so the effects would have to be investigated.

While this special casing doesn't look good, I think it will result in pretty good gains for Razor semantic tokens which imo might be an appropriate trade-off since the Razor LSP editor is on by default for users.

CyrusNajmabadi commented on src/Features/Core/Portable/SemanticClassificationCache/SemanticClassificationCacheUtilities.cs:
```diff
@@ -30,5 +36,71 @@ internal static class SemanticClassificationCacheUtilities
+            if (isRazorDoc)
+            {
+                var spans = await Classifier.GetClassifiedSpansAsync(document, textSpan, cancellationToken).ConfigureAwait(false);
+                classifiedSpans.AddRange(spans);
```
hrmm... should this code just move up to the razor caller instead?

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -162,6 +166,32 @@ static SemanticTokensHelpers()
+            // casing: https://github.com/dotnet/razor-tooling/issues/5850
+            // Until then, the `AddSemanticClassificationsAsync` call below is special-cased to deal with both scenarios.
+            await SemanticClassificationCacheUtilities.AddSemanticClassificationsAsync(
+                document, textSpan, classificationService, classifiedSpans, isRazorDoc, cancellationToken).ConfigureAwait(false);
```
so instead of passing isRazorDoc here (and then having AddSemandticClassificationsAsync do odd stuff (like adding syntactic classifications), i would instead have the code *here* fork it's behavior (with the good comment you have), and say htat in the razor case we get all diagnostics, where for all otehrs we just get semantic classification.

CyrusNajmabadi commented on src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs:
```diff
@@ -49,6 +62,12 @@ public override async Task AnalyzeDocumentAsync(Document document, SyntaxNode bo
 
                 var solution = document.Project.Solution;
+
+                // Only cache classifications in non-LSP scenarios. LSP caching is handled separately.
```
@dibarbet here as well...

interesting.  why is this @allisonchou ?  why not have the normal incremental analyzer handle this.  note: if this is the right way to go, then this comment needs beefing up.  Specifically, this is an example of a comment htat just repeats what the code is doing but doesn't provide clarity as to 'why'.  I cannot look at this code/comment and say: ah, yes, that makes sense, we def should be doing things in this fashion.

allisonchou replied:
The main issue with using the default logic is the method's call to [`document.IsOpen()`](https://github.com/dotnet/roslyn/blob/498d5b1f6feace039534035a08aa864c0cd26d1e/src/VisualStudio/Core/Def/Implementation/SemanticClassificationCache/SemanticClassificationCacheIncrementalAnalyzerProvider.cs#L48). This call is problematic in LSP for two reasons:

1) It bases whether the document is open on the non-LSP workspace rather than the LSP workspace.
2) Razor back-end generated documents won't be considered open even if the front-end document is open. This means we would have to special case Razor (via exposing a new document service, workspace service or something else).

Putting the caching logic in the LSP handler addresses both 1) and 2) by ensuring we only cache open files in the LSP workspace, since we only get semantic token requests for the file(s) currently open. In an ideal world, I think we'd have the `IsOpen` check correspond to the LSP workspace and that workspace can then specially check for open Razor documents, but I believe that is a much bigger work item.

allisonchou commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:
```diff
@@ -45,15 +48,44 @@ public SemanticTokensRangeHandler()
+            await client.TryInvokeAsync<IRemoteSemanticClassificationCacheService>(
+                document.Project,
+                (service, solutionInfo, cancellationToken) => service.CacheSemanticClassificationsAsync(solutionInfo, document.Id, cancellationToken),
+                cancellationToken).ConfigureAwait(false);
```
(Same explanation as https://github.com/dotnet/roslyn/pull/58238#discussion_r768215940)

allisonchou commented:
@CyrusNajmabadi whenever you have time, this should be ready to review again. I've addressed your feedback + responded to the changes from your other PRs.

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
+
+            // Classified spans are not guaranteed to be returned in a certain order so we sort them to be safe.
+            classifiedSpans.Sort(ClassifiedSpanComparer.Instance);
+        }
```
nice.  this is all much cleaner IMO :)

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -371,5 +406,12 @@ private static int GetTokenTypeIndex(string classificationType, Dictionary<strin
+            public static readonly ClassifiedSpanComparer Instance = new();
+
+            public int Compare(ClassifiedSpan x, ClassifiedSpan y) => x.TextSpan.CompareTo(y.TextSpan);
+        }
```
note: as per previous discussion, i don't think this is necessary.  but if you want to keep it, that's fine.

allisonchou replied:
Interestingly enough, `GetClassifiedSpansAsync` and `AddSemanticClassificationsAsync` each return classifications in opposite order from one another. I think we might want to keep the check just in case.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:
```diff
@@ -92,13 +103,21 @@ private static void RemoveAdditiveSpans(ArrayBuilder<ClassifiedSpan> spans)
+            else
+            {
+                filledInSyntaxSpans = syntaxSpans;
+                filledInSemanticSpans = semanticSpans;
```
i don't love this.  primarily because you're overiwring a variable that feels like it is controlled by the using block.  Technically, this works... it's just oogy :)

allisonchou replied:
Gotcha. I slightly refactored this portion in my latest push just now, to me it looks a little cleaner but let me know if you still have any concerns.

allisonchou replied:
Ok, I actually had to revert my refactoring - guess the logic wasn't equivalent 😦 lmk if you have any other suggestions

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:
```diff
@@ -169,15 +188,15 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
                 // See if we can find a semantic part to replace this syntax part.
                 var replacementIndex = semanticParts.FindIndex(
-                    lastReplacementIndex, t => t.TextSpan == syntaxPartAndSpan.TextSpan);
+                    lastReplacementIndex, t => t.TextSpan.OverlapsWith(syntaxPartAndSpan.TextSpan));
```
why this change?

allisonchou replied:
I believe this is an existing bug in the logic related to the comment I added below:

```
                    // There may be multiple semantic parts corresponding to a single
                    // syntactic part, so we might need to go through a syntactic part
                    // multiple times to verify. For example, this is the case with
                    // verbatim string literals containing string escape characters.
```

Essentially, there's a chance the TextSpan of the semantic part will not exactly match the TextSpan of the syntactic part. This occurs with string literals containing string escape characters. The string escape character will not have the same exact span as the syntactic part and therefore will not be considered. I believe instead calling `OverlapsWith` mitigates the problem.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:
```diff
@@ -189,14 +208,20 @@ private static void AdjustSpans(ArrayBuilder<ClassifiedSpan> spans, TextSpan wid
+                    // syntactic part, so we might need to go through a syntactic part
+                    // multiple times to verify. For example, this is the case with
+                    // verbatim string literals containing string escape characters.
+                    i--;
```
thanks for the example.  do we have ests for this?

allisonchou replied:
Yep, a test in `SemanticTokensRangeTests` actually caught this error 😄

CyrusNajmabadi replied:
ok great!

CyrusNajmabadi replied:
actually, can you ensure we have a few tests here.  for example, if a verbatim literal contains multiple escape characters like so:

`$"foo""bar""baz{x + y}a""b""c"`

something to stress teh system

allisonchou replied:
I added a regex test in my latest push which I believe accomplishes this 🙂

CyrusNajmabadi commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
+            {
+                var spans = await ClassifierHelper.GetClassifiedSpansAsync(
+                    document, textSpan, cancellationToken, removeAdditiveSpans: false, fillInClassifiedSpanGaps: false).ConfigureAwait(false);
+                classifiedSpans.AddRange(spans);
```
so, i have a subjective question for you.   It was def me that pushed using ClassifierHelper.getClassifiedSpansAsync.  However, in order to do so you had to add a fair amount of logic to it to get what you need for the LSP case.

My overall question is which do *you* think is better of the following two options:

1. keep the PR as is, and have the extra complexity in that method.
2. just provide a totally different method that, while similar to the above, is independent and dedicated for the LSP case.

If you feel like '2' actually ends up adding/repeating too much code.  Then let's stick with what's in the PR. However, if you feel that the two methods are different enough that trying to force them to be one method (along with several booleans controlling) is just to overwrought, then maybe we should just have two.

I don't have enough context to say from reading this.  However, my hope was that you could just use hte method as is.  the need to pass all this customization logic in is making me rethink that :)

allisonchou replied:
Just tested out option 2). Unfortunately, it does feel like a substantial amount of the code is duplicated between the LSP and non-LSP methods. The only portions that differ are a) whether to remove additive spans and b) whether to fill in classified span gaps. The optional parameters don't feel great to me either, but I feel like the amount of duplication we avoid may be worth the tradeoff.

CyrusNajmabadi replied:
sgtm.

allisonchou commented:
@CyrusNajmabadi thanks for your review! I've responded to your feedback. Would you mind taking a look again whenever you have time?

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
+            Document document,
+            TextSpan textSpan,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            bool isRazorDoc,
```
more of a naming nit - perhaps we could rename this to be something like `includeSyntacticClassifications`

ryanbrandenburg commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -149,9 +150,8 @@ static SemanticTokensHelpers()
-            var classifiedSpans = Classifier.GetClassifiedSpans(document.Project.Solution.Workspace.Services, semanticModel, textSpan, options, cancellationToken);
-            Contract.ThrowIfNull(classifiedSpans, "classifiedSpans is null");
+            using var _ = ArrayBuilder<ClassifiedSpan>.GetInstance(out var classifiedSpans);
+            await GetClassifiedSpansForDocumentAsync(document, textSpan, classifiedSpans, isRazorDoc, cancellationToken).ConfigureAwait(false);
```
Not a fan of this kind of "secret return" inside of the `classifiedSpans`. Could `GetClassifiedSpansForDocumentAsync` return `Task<ClassifiedSpan[]>` so we can be explicit about what had work done on it? Also, if `ArrayBuilder<ClassifiedSpan>` is static couldn't we avoid passing it and just create the instance inside of `GetClassifiedSpansForDocumentAsync`?

CyrusNajmabadi replied:
note: in terms of naming, we try to do 'Get' to have a return value, and 'Add' if someone will pass in the collection to fill.

ryanbrandenburg commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensRangeHandler.cs:
```diff
@@ -50,8 +50,8 @@ public SemanticTokensRangeHandler()
             var (tokensData, isFinalized) = await SemanticTokensHelpers.ComputeSemanticTokensDataAsync(
-                context.Document, SemanticTokensHelpers.TokenTypeToIndex,
-                request.Range, cancellationToken).ConfigureAwait(false);
+                context.Document, SemanticTokensHelpers.TokenTypeToIndex, request.Range,
```
Nit: Personally I'm of the opinion that parameters should be all on the same line or each on their own line to avoid diffs like this.

ryanbrandenburg commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:
```diff
@@ -17,7 +19,7 @@ namespace Microsoft.CodeAnalysis.LanguageServer.UnitTests.SemanticTokens
     {
         [Fact]
-        public async Task TestGetSemanticTokensRange_FullDocAsync()
+        public async Task TestGetSemanticTokensRange_FullDoc()
```
Why change the name? Seems like that'd mess up statistics if nothing else.

ryanbrandenburg reviewed (commented):
Only complaint is the "secret return", but you probably want to wait for review from someone with more expertise in Roslyn.

dibarbet commented on src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs:
```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
+            if (isRazorDoc)
+            {
+                var spans = await ClassifierHelper.GetClassifiedSpansAsync(
+                    document, textSpan, cancellationToken, removeAdditiveSpans: false, fillInClassifiedSpanGaps: false).ConfigureAwait(false);
```
might be good to make a note of why the flags are set this way for LSP.  I saw a mention below for removeAdditiveSpans, but not for fillInClassifiedSpanGaps

dibarbet commented on src/Features/LanguageServer/ProtocolUnitTests/SemanticTokens/SemanticTokensRangeTests.cs:
```diff
@@ -28,41 +30,36 @@ static class C { }
         }
 
         [Fact]
-        public async Task TestGetSemanticTokensRange_PartialDocAsync()
```
did this test go missing?  I'm not seeing a new test that verifies the partial (1,0) -> (2,0) positions

allisonchou replied:
Whoops, you're totally right. Thanks! Added it back.

CyrusNajmabadi commented on src/Workspaces/Core/Portable/Classification/ClassifierHelper.cs:
```diff
@@ -90,19 +101,25 @@ private static void RemoveAdditiveSpans(ArrayBuilder<ClassifiedSpan> spans)
+                FillInClassifiedSpanGaps(widenedSpan.Start, semanticSpans, filledInSemanticSpans);
+
+                syntaxSpans = filledInSyntaxSpans;
+                semanticSpans = filledInSemanticSpans;
```
this is *definitely* not going to work.  :)  remember that when the using goes out of scope, we'll put the array back in teh pool.  meaning someone else might grab on it and party on it at the same time as us.  i have a recommendtation on how to structure this code.  ping me :)

CyrusNajmabadi reviewed (approved):
Thanks!

allisonchou commented:
Thanks everyone!
